adapters/celestia/src/verifier/mod.rs

```diff
@@ -29,12 +33,29 @@ pub const PARITY_SHARES_NAMESPACE: NamespaceId = NamespaceId(hex_literal::hex!("
 impl BlobTransactionTrait for BlobWithSender {
     type Data = BlobIterator;
     type Address = CelestiaAddress;
+
     fn sender(&self) -> CelestiaAddress {
         self.sender.clone()
     }
 
-    fn data(&self) -> Self::Data {
-        self.blob.clone().into_iter()
+    fn data(&self) -> BufWithCounter<Self::Data> {
+        BufWithCounter::new(self.blob.clone().into_iter())
+    }
+
+    fn hash(&self) -> [u8; 32] {
```

Comment:
In the current API the `blob` will be cloned in a call to `hash` and `data`. Isn't always to be the same caller for these two methods? If the `data()` and `hash()` are called only once in the same place then we could unify them and save one clone.